Enable Jupyter kernels for batch execution

TuturBaba opened the pull request:
Hello,
I am a master's student doing my internship with @yvanlebras . I need to create a workflow on Galaxy and noticed an issue with the latest Jupyter tool in Galaxy. Since it cannot be used in batch mode, I looked for a simple way to fix it. From what I implemented, it seems to work. I used pip install inside Conda and, most importantly, ipykernel, which allows Jupyter to correctly recognize the kernel paths and manually name them.

TuturBaba committed:
Enable Jupyter kernels for batch execution

bgruening commented:
Thanks a lot @TuturBaba.

Is your hypothesis that the conda package of ipykernel is broken? Or just too old?

bgruening commented on Dockerfile:
```diff
@@ -35,30 +35,45 @@ RUN conda install --yes \
-        'r-tidyverse' \
-        'unixodbc' && \
+    conda create -n ansible-kernel --yes && \
+    conda run -n ansible-kernel pip install ipykernel bioblend galaxy-ie-helpers && \
```
why do you install the ipykernel kernel in the ansible env?

TuturBaba replied:
I just noticed that there are quite a few mistakes and that I didn’t fully understand what I had done. So, I’m reviewing my work and will get back to it later.
The main issue is with the metadata, as shown in the image. Since I focused on R and Python, I saw that I could modify the metadata using ipkernel. Without thinking too much, I applied the same method to other languages, but this only changed the name while the language remained Python.
So, I need to find a solution to ensure that when running jupyter kernelspec list, the different kernels are correctly configured. At the moment, with the current setup, only the kernels that come with the base image work: python3 and julia-1.10
![Screenshot from 2025-03-25 14-39-05](https://github.com/user-attachments/assets/179db8c0-f44f-4216-8991-97eebf246d6a)

TuturBaba replied:
And I forgot, but even the kernel that come with base image don't work by batch

bgruening replied:
Isn't that a problem https://github.com/anaconda/nb_conda_kernels tries to solve?

TuturBaba replied:
Maybe, but I haven't found a complete solution, however, I used this

TuturBaba committed:
Update Dockerfile

TuturBaba committed:
Update Dockerfile

TuturBaba committed:
Update Dockerfile

TuturBaba commented:
Hello,
I have modified my Dockerfile, and I think I understand how it works, so I should be able to answer your questions.
However, I have one question: What is better, a single long RUN command or multiple RUN commands? I know that using multiple RUN commands can make the Dockerfile easier to manage since you don’t have to rebuild everything from scratch if something changes. But in terms of optimization, what is the best approach?

bgruening commented:
One RUN command is recommended or you need to use a multi-stage Docker build. This is all to reduce the size of the final image.

TuturBaba committed:
Update Dockerfile

TuturBaba committed:
Update Dockerfile

TuturBaba committed:
Update Dockerfile

bgruening commented on Dockerfile:
```diff
@@ -35,33 +35,41 @@ RUN conda install --yes \
+	    'unixodbc' && \
+    conda run -n rlang-kernel R -e 'IRkernel::installspec(user = TRUE)' && \
+    conda create -n bash-kernel --yes bash_kernel bioblend galaxy-ie-helpers && \
+    conda run -n bash-kernel python -m bash_kernel.install --user && \
```
you don't need to set the --disply-name here?

bgruening replied:
or the --name?

TuturBaba replied:
It's possible to use --user or --display-name, though it's not certain. However, it works without them. Using .install changes the metadata

bgruening replied:
So what happens if I want to batch run a bash script is that now possible?

TuturBaba replied:
Normally, yes. To test my Dockerfile, I add this line at the end:  
ADD ./bash.ipynb /home/$NB_USER/work/bash.ipynb

Then, I run the following command:  
docker run --rm -it -v $(pwd):/home/jovyan/work test bash

inside, I execute:  
jupyter nbconvert --to notebook --execute /home/jovyan/work/bash.ipynb --output /home/jovyan/work/bash_output.ipynb

In the output, I get "3". If it's indeed batch processing, then it works.

And inside bash.ipynb, there's just:  
a=$((1+2))  
echo $a

TuturBaba replied:
Normally, I tested python and R, and they work. Octave doesn't work, even when I using jupyter notebook in galaxy, it doesn't work. Ansible works, but I don't know much about it.

TuturBaba replied:
I found other issues in what I did while trying to use it more. For example, when I trying to install something for R, I usually use system("conda install r-terra"), but it doesn't work, same for `install.packages()`. However, I have some ideas on how to fix these issues.

TuturBaba committed:
Update Dockerfile

bgruening commented on Dockerfile:
```diff
@@ -35,33 +35,41 @@ RUN conda install --yes \
+	    'r-tidymodels' \
+	    'r-tidyverse' \
+	    'unixodbc' && \
+    conda run -n rlang-kernel R -e 'IRkernel::installspec(user = TRUE)' && \
```
You can also add here something like `(name = 'rlang-kernel', displayname = 'R')`

bgruening replied:
and I guess for the other kernels as well

TuturBaba replied:
Yes u can : `conda run -n rlang-kernel R -e "IRkernel::installspec(user = TRUE, name = 'rlang-kernel', displayname = 'R')"`
But it's also possible without specifying a name: `conda run -n python-kernel-3.12 python -m ipykernel install --user`

TuturBaba commented:
I have a question, Jupyter version 1.0.0 does not use a Conda envs. So I was wondering why you chose to create environments. I did a quick test, and it seems to work fine (without any complications).

TuturBaba committed:
Configuration to .Rprofile

Before this change, there was an issue with the Dockerfile when trying to install R libraries using commands like system("conda install r-terra"). While the libraries were technically installed, they were not placed in the correct location.  I confirmed this by checking with .libPaths().

After discovering the problem, I searched for a solution and found that by adding a .Rprofile file with the correct paths, the installation now works as expected. This modification sets the CONDA_PREFIX and updates the PATH in the .Rprofile to ensure R libraries are installed in the appropriate environment

bgruening commented:
Because people wanted to install new packages for R, python, c++ whatever. The only package manager that enables all this and does not waste resources on compling is conda (in our community)

TuturBaba commented:
I understand the benefit of using Conda to manage dependencies, but why create a new environment (`conda create`) instead of installing packages directly in the base environment with `conda install`? Is there a specific advantage to isolating installations?  

PS: By the way, I made a commit that fixes my issue, so I think this version is working fine.

bgruening commented:
If you mix many different languages and many different packages my assumption was that if you install new packages over time it gets more complicated and brittle. Maybe I'm wrong and this is overengeenered. Is now everything working for you? Can you maybe outline how you tested this locally and maybe add it to the readme?

bgruening reviewed (approved):
Will merge in the next days if no one else has comments and release a new version.

yvanlebras commented:
Thank you Björn! I think the PR is not ready to merge for now, the issue for R is solved, but not for Python (in batch mode, when a conda package is installed, it works fine BUT when the code is trying to use the installed conda package, Arthur has an error as the PATH is not the good one). Arthur is trying to look at it more deeply and try to apply the best patch he can if possible for all envs.

TuturBaba commented:
Hello Bjorn,
I have a script that seems correct, and I’m currently testing it to see if it works as expected.
I wanted to know if you have any test scripts available on your side, especially for Ansible and Octave, as I don’t have any knowledge of how they work.

bgruening commented:
Ok, just checking, there are no new commits, the current version is working for you?

TuturBaba committed:
Improve kernel installation method and remove nb_conda_kernels

Hello, I’m going to explain my work during the testing phase. First, I use Conda packages for different languages. I use the python -m command to access certain package options. When 
I use this method to install the kernel, I don’t need nb_conda_kernels.

So, I can remove nb_conda_kernels, and this allows the base image kernels (Julia and Python) to get the correct path and metadata again. That’s why I didn’t create a new Python environment, but I can create one if needed. However, like with R, the path won’t be correct. I think I can define a solution to fix that if you want.

To test my Dockerfile, I use three scripts, which you can find on my GitHub: https://github.com/TuturBaba/Antarctic/tree/main/script_test_dockerfile . I only test R, Python, and Bash because I don’t have experience with other languages.I used basic commands, like doing math operations, writing to a file, and installing packages.

To test this, I simply added the following line at the end of my Dockerfile: ADD ./R.ipynb /import/R.ipynb
 Then I run:  RUN chmod -R 777 /import

TuturBaba committed:
ADD PATH in .Rprofile

There was a problem with "proj", the path after download was not the right one. Because of the way we manage R, it means that we can find other problems like

yvanlebras commented:
Hi Arthur, Björn. Seems to me this is ok to merge isn't it ? For links, related issues are https://github.com/usegalaxy-eu/galaxy/issues/244 and https://github.com/usegalaxy-eu/galaxy/issues/266

bgruening commented:
I will create a new release and a new Galaxy tool version. Thanks everyone!

bgruening commented:
Ok, it is deployed: https://usegalaxy.eu/root?tool_id=interactive_tool_jupyter_notebook make sure to use the latest 1.0.2 version.

bgruening commented:
@TuturBaba it seems this has broken some other functionality. Can you try to create a new env with some new packages? I think its not possible to change into this new env anymore via the UI.

![grafik](https://github.com/user-attachments/assets/4b4895e8-ecfa-43e2-8489-4369bbcb76af)

Can you maybe create a small bash script that is testing your usecase, independent of Galaxy, just this container and a notebook? That would help me I guess to make sure your use-case is working but still not breaking the venv selector.

TuturBaba commented:
@bgruening I didn’t fully understand what you want to do.
On my side, I tried to create a kernel directly in the Galaxy Jupyter tool, and I was able to create a new Python kernel with these commands (in terminal):

conda create -n test-kernel python=3.10 ipykernel -y
conda activate test-kernel
python -m ipykernel install --user --name test-kernel --display-name "Python (test)"

Also, running without Galaxy works for me because I use the container directly on my computer.
But if you need something specific, please give more details.
